include/nektar_interface/function_basis_projection_alt.hpp

```diff
@@ -0,0 +1,179 @@
+                               // something process multiple componants at once
+                               // wasteful to do one at a time probably
+          V &global_coeffs, bool force_thread_per_dof = false) {
+
+    this->dh_global_coeffs.realloc_no_copy(global_coeffs.size());
+    fill_device_buffer_and_wait(this->dh_global_coeffs.d_buffer.ptr, U(0.0),
+                                global_coeffs.size(), this->sycl_target->queue);
+
+    if (this->mesh->get_ndim() == 2) {
+      if (this->sycl_target->queue.get_device().is_gpu() ||
+          force_thread_per_dof) {
+        project_inner<Project::eQuad<Project::ThreadPerDof>, U>(particle_group,
+                                                                sym, component)
+            .wait();
+        project_inner<Project::eTriangle<Project::ThreadPerDof>, U>(
+            particle_group, sym, component)
+            .wait();
+      } else {
+        project_inner<Project::eQuad<Project::ThreadPerCell>, U>(particle_group,
+                                                                 sym, component)
+            .wait();
+
+        project_inner<Project::eTriangle<Project::ThreadPerCell>, U>(
+            particle_group, sym, component)
+            .wait();
+      }
+    } else {
+      if (this->sycl_target->queue.get_device().is_gpu() ||
+          force_thread_per_dof) {
+        project_inner<Project::eHex<Project::ThreadPerDof>, U>(particle_group,
+                                                               sym, component)
+            .wait();
+        project_inner<Project::ePrism<Project::ThreadPerDof>, U>(particle_group,
+                                                                 sym, component)
+            .wait();
+        project_inner<Project::eTet<Project::ThreadPerDof>, U>(particle_group,
+                                                               sym, component)
+            .wait();
+        project_inner<Project::ePyramid<Project::ThreadPerDof>, U>(
+            particle_group, sym, component)
+            .wait();
+      } else {
+        project_inner<Project::eHex<Project::ThreadPerCell>, U>(particle_group,
+                                                                sym, component)
+            .wait();
+        project_inner<Project::ePrism<Project::ThreadPerCell>, U>(
+            particle_group, sym, component)
+            .wait();
+        project_inner<Project::eTet<Project::ThreadPerCell>, U>(particle_group,
+                                                                sym, component)
+            .wait();
+        project_inner<Project::ePyramid<Project::ThreadPerCell>, U>(
+            particle_group, sym, component)
+            .wait();
+      }
+    }
+
+    // copyback
+    this->sycl_target->queue
+        .memcpy(global_coeffs.begin(), this->dh_global_coeffs.d_buffer.ptr,
+                global_coeffs.size() * sizeof(U))
+        .wait();
+  }
+};
+} // namespace NESO
```

Comment:
End include guard from comment at top of file.
```suggestion
} // namespace NESO
#endif
```

Reply:
as above